src/Aquifer.Public.API/Endpoints/Resources/Updates/List/Endpoint.cs

```diff
@@ -16,23 +18,70 @@ public override void Configure()
             s.Description =
-                "For a given UTC timestamp and optional language id, get a list of resource ids that are new or have been updated since the provided timestamp. This is intended for users who are storing Aquifer data locally and want to fetch new content.";
+                "For a given UTC timestamp, get a list of resource ids that are new or have been updated since the provided timestamp. This is intended for users who are storing Aquifer data locally and want to fetch new content.";
         });
     }
 
     public override async Task HandleAsync(Request req, CancellationToken ct)
     {
-        Response.Items = await dbContext.ResourceContentVersions
-            .Where(x => (!req.LanguageId.HasValue || x.ResourceContent.LanguageId == req.LanguageId) &&
-                x.IsPublished &&
-                x.Updated >= req.Timestamp)
+        var query = await GetQuery(req, ct);
+        var totalCount = await GetTotalResourceCountAsync(req, query, ct);
+
+        if (totalCount == 0)
+        {
+            await SendOkAsync(new Response(), ct);
+            return;
+        }
+
+        var items = await GetResourceUpdatesAsync(req, query, ct);
+        Response = new Response
+        {
+            Items = items,
+            TotalItemCount = totalCount,
+            Offset = req.Offset
+        };
+    }
+
+    private async Task<List<ResponseContent>> GetResourceUpdatesAsync(Request req, IQueryable<ResourceContentVersionEntity> query,
+        CancellationToken ct)
+    {
+        var languageCodeByIdMap = await _cachingLanguageService.GetLanguageCodeByIdMapAsync(ct);
+        return await query
+            .OrderByDescending(r => r.Updated)
+            .Skip(req.Offset)
+            .Take(req.Limit)
             .Select(x => new ResponseContent
             {
                 ResourceId = x.ResourceContentId,
                 LanguageId = x.ResourceContent.LanguageId,
+                LanguageCode = languageCodeByIdMap[x.ResourceContent.LanguageId],
                 UpdateType = x.Version == 1 ? ResponseContentUpdateType.New : ResponseContentUpdateType.Updated,
                 Timestamp = x.Updated
             })
             .ToListAsync(ct);
     }
+
+    private async Task<int> GetTotalResourceCountAsync(Request req, IQueryable<ResourceContentVersionEntity> query, CancellationToken ct)
+    {
+        var totalCount = await query.CountAsync(ct);
+        if (req.Offset >= totalCount && req.Offset != 0)
+        {
+            ThrowError($"Offset of {req.Offset} equals or exceeds total of {totalCount}");
+        }
+
+        return totalCount;
+    }
+
+    private async Task<IQueryable<ResourceContentVersionEntity>> GetQuery(Request req, CancellationToken ct)
+    {
+        var languageId = req.LanguageCode is not null
+            ? await _cachingLanguageService.GetLanguageIdAsync(req.LanguageCode, ct) ?? 0
+            : req.LanguageId;
```

Comment:
I wrote a utility method [here](https://github.com/BiblioNexusStudio/aquifer-server/blob/6007e86325e95b72c915f208d711b77ff78fb060/src/Aquifer.Common/Services/Caching/CachingLanguageService.cs#L82-L83) that is useful for the Public API where both `languageCode` and `languageId` can be passed.  I suggest using it here like this:
```
var (isValidLanguageId, isValidLanguageCode, validLanguageId) = await ValidateLanguageIdOrCodeAsync(req.LanguageId, req.LanguageCode, shouldRequireInput: false, ct);
```

This will validate that either `languageId` or `languageCode` have good data or are both `null` (when `shouldRequireInput` is `false` and will return a `languageId` that is guaranteed to exist (or possibly `null` when `shouldRequireInput` is `false` which is a valid state for this endpoint).

Thus, you could return a bad request if either `isValidLanguageId` or `isValidLanguageCode` is `false` (with a good error message for either case that includes the invalid request parameter) and then use the `int? validLanguageId` in the query below (replacing the current `languageId` variable).

Reply:
nifty. thanks!